test/Test/Data/Mutable/Set.hs

```diff
@@ -129,9 +141,39 @@ memberDelete2Test val1 val2 set = fromRead (Set.member val2 set)
         memberVal2
         (getFst Linear.$ Set.member val2 (Set.delete val1 set))
 
+memberUnion :: Property
+memberUnion = property $ do
+  l <- forAll list
+  l' <- forAll list
+  v <- forAll value
+  let valInEither = elem v l || elem v l'
+  test $ unur Linear.$
+    Set.fromList l
+      (\s -> Set.fromList l'
+        (\s' -> memberUnionTest v valInEither (Set.union s s')))
+
+memberUnionTest :: Int -> Bool -> Set.Set Int %1-> Ur (TestT IO ())
+memberUnionTest val result union = Set.member val union Linear.&
+  \(Ur b, s') -> lseq s' (Ur (b === result))
+
+memberIntersection :: Property
+memberIntersection = property $ do
+  l <- forAll list
+  l' <- forAll list
+  v <- forAll value
+  let valInBoth = elem v l && elem v l'
+  test $ unur Linear.$
+    Set.fromList l
+      (\s -> Set.fromList l'
+        (\s' -> memberIntersectTest v valInBoth (Set.intersection s s')))
+
+memberIntersectTest :: Int -> Bool -> Set.Set Int %1-> Ur (TestT IO ())
+memberIntersectTest val result inter = Set.member val inter Linear.&
+  \(Ur b, s') -> lseq s' (Ur (b === result))
```

Comment:
This test is correct. But I contend that it doesn't test much, as I assume that, almost always, `v` will not be a member of either `l` or `l'` (let alone both). Therefore, this test will almost always be testing that `False === False`.

Instead, I'd suggest writing a `toList` function of sort, and testing that the output list is the expected one. Individual tests will be a bit more expensive, but much more informative.

Reply:
Ah, so pseudo-homomorphism tests. It's not trivial to see, but these are sufficient to replace defining axioms. Done below.